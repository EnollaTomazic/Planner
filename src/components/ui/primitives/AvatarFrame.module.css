.root {
  --avatar-size-sm: clamp(
    calc(var(--space-6) + var(--space-1)),
    calc(var(--space-6) + var(--space-1) + 2vw),
    calc(var(--space-7) + var(--space-2))
  );
  --avatar-size-md: clamp(
    calc(var(--space-7)),
    calc(var(--space-7) + var(--space-2) + 3vw),
    calc(var(--space-8) + var(--space-1))
  );
  --avatar-size-lg: clamp(
    calc(var(--space-7) + var(--space-1)),
    calc(var(--space-7) + var(--space-1) + 4vw),
    calc(var(--space-8) + var(--space-4))
  );
  --avatar-size: var(--avatar-size-md);
  --avatar-rim: calc(var(--space-2) / 2);
  --avatar-surface: hsl(var(--card) / 0.92);
  --avatar-outline: 0 0 0 var(--hairline-w) hsl(var(--card-hairline) / 0.65);
  --avatar-rim-gradient: conic-gradient(
    from 140deg at 50% 50%,
    hsl(var(--accent) / 0.88),
    hsl(var(--accent-2) / 0.9),
    hsl(var(--ring) / 0.85),
    hsl(var(--accent) / 0.88)
  );
  --avatar-glow-gradient: radial-gradient(
    circle at 50% 32%,
    hsl(var(--lav-deep) / 0.55),
    transparent 70%
  );
  --avatar-glow-mask: radial-gradient(
    circle at 50% 32%,
    hsl(var(--shadow-color) / 0.95) 24%,
    hsl(var(--shadow-color) / 0.35) 60%,
    transparent 78%
  );
  --avatar-glow-blur: var(
    --hero-illustration-blur,
    calc(var(--space-4) + var(--space-1))
  );
  --portrait-size: var(--avatar-size);
  --portrait-rim: var(--avatar-rim);
  --portrait-glow: var(--avatar-glow-blur);
}

.sizeSm {
  --avatar-size: var(--avatar-size-sm);
}

.sizeMd {
  --avatar-size: var(--avatar-size-md);
}

.sizeLg {
  --avatar-size: var(--avatar-size-lg);
}

.framelessSurface {
  width: var(--avatar-size);
  height: var(--avatar-size);
  background: var(--avatar-surface);
  box-shadow: var(--avatar-outline);
}

.rim {
  width: calc(var(--avatar-size) + var(--avatar-rim) * 2);
  height: calc(var(--avatar-size) + var(--avatar-rim) * 2);
  padding: var(--avatar-rim);
  background: var(--avatar-rim-gradient);
}

.inner {
  width: var(--avatar-size);
  height: var(--avatar-size);
  background: var(--avatar-surface);
  box-shadow: var(--avatar-outline);
}

.glow {
  background: var(--avatar-glow-gradient);
  mask-image: var(--avatar-glow-mask);
  -webkit-mask-image: var(--avatar-glow-mask);
}

.media {
  object-position: center 20%;
}
