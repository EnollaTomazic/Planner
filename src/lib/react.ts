import * as React from "react";

export const hasTextContent = (node: React.ReactNode): boolean => {
  if (node === null || node === undefined) return false;
  if (typeof node === "boolean") return false;
  if (typeof node === "string") return node.trim().length > 0;
  if (typeof node === "number") return true;
  if (Array.isArray(node)) return node.some((item) => hasTextContent(item));
  if (React.isValidElement<{ children?: React.ReactNode }>(node)) {
    return hasTextContent(node.props.children ?? null);
  }
  return false;
};

export function observeMediaQuery(
  query: string,
  onChange: (matches: boolean, mediaQuery: MediaQueryList) => void,
): (() => void) | undefined {
  if (typeof window === "undefined" || typeof window.matchMedia !== "function") {
    return undefined;
  }

  let mediaQuery: MediaQueryList;

  try {
    mediaQuery = window.matchMedia(query);
  } catch {
    return undefined;
  }

  const handleChange = (event: MediaQueryListEvent): void => {
    onChange(event.matches, mediaQuery);
  };

  const handleLegacyChange = (): void => {
    onChange(mediaQuery.matches, mediaQuery);
  };

  onChange(mediaQuery.matches, mediaQuery);

  if (typeof mediaQuery.addEventListener === "function") {
    mediaQuery.addEventListener("change", handleChange);
    return () => {
      mediaQuery.removeEventListener("change", handleChange);
    };
  }

  if (typeof mediaQuery.addListener === "function") {
    mediaQuery.addListener(handleLegacyChange);
    return () => {
      mediaQuery.removeListener(handleLegacyChange);
    };
  }

  return undefined;
}

export function useMatchMedia(
  query: string,
  getServerSnapshot?: () => boolean,
): boolean {
  const getMatches = React.useCallback(() => {
    if (typeof window === "undefined") {
      return typeof getServerSnapshot === "function" ? getServerSnapshot() : false;
    }

    if (typeof window.matchMedia !== "function") {
      return false;
    }

    try {
      return window.matchMedia(query).matches;
    } catch {
      return false;
    }
  }, [query, getServerSnapshot]);

  const [matches, setMatches] = React.useState<boolean>(getMatches);

  React.useEffect(() => {
    return observeMediaQuery(query, (nextMatches) => {
      setMatches(nextMatches);
    });
  }, [query]);

  return matches;
}
