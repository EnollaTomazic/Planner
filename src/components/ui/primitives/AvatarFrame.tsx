import * as React from "react";
import type { ImageProps } from "next/image";

import { cn } from "@/lib/utils";

import styles from "./AvatarFrame.module.css";

type AvatarFrameSize = "sm" | "md" | "lg";

type MediaElement = React.ReactElement<
  ImageProps | React.ImgHTMLAttributes<HTMLImageElement>
>;

type AvatarFrameOwnProps<T extends React.ElementType = "div"> = {
  as?: T;
  frame?: boolean;
  size?: AvatarFrameSize;
  media?: MediaElement;
  before?: React.ReactNode;
  after?: React.ReactNode;
  children?: React.ReactNode;
  innerClassName?: string;
  surfaceClassName?: string;
  rimClassName?: string;
  glowClassName?: string;
  railClassName?: string;
};

export type AvatarFrameProps<T extends React.ElementType = "div"> =
  AvatarFrameOwnProps<T> &
    Omit<React.ComponentPropsWithoutRef<T>, keyof AvatarFrameOwnProps<T>>;

const sizeClassNames: Record<AvatarFrameSize, string> = {
  sm: styles.sizeSm,
  md: styles.sizeMd,
  lg: styles.sizeLg,
};

function coerceMedia(media?: MediaElement) {
  if (!media || !React.isValidElement(media)) {
    return media;
  }

  return React.cloneElement(media, {
    className: cn(
      "h-full w-full object-contain object-[center_20%]",
      styles.media,
      media.props.className,
    ),
  });
}

function AvatarFrame<T extends React.ElementType = "div">({
  as,
  frame = true,
  size = "md",
  media,
  before,
  after,
  children,
  className,
  innerClassName,
  surfaceClassName,
  rimClassName,
  glowClassName,
  railClassName,
  ...rest
}: AvatarFrameProps<T>) {
  const Comp = (as ?? "div") as React.ElementType;
  const processedMedia = coerceMedia(media);

  return (
    <Comp
      className={cn(
        "relative isolate flex shrink-0 items-center justify-center",
        styles.root,
        sizeClassNames[size],
        className,
      )}
      {...rest}
    >
      {before}
      {frame ? (
        <>
          <span
            aria-hidden
            className={cn(
              "pointer-events-none absolute -inset-[calc(var(--portrait-rim)*1.6)] rounded-full blur-[var(--hero-illustration-blur,var(--portrait-glow))] opacity-[var(--hero-illustration-opacity,0.8)]",
              styles.glow,
              glowClassName,
            )}
          />
          <span
            aria-hidden
            className={cn(
              "glitch-rail pointer-events-none absolute -inset-[calc(var(--portrait-rim)*1.1)] rounded-full mix-blend-screen opacity-[var(--glitch-intensity-hero-rail,0.74)]",
              railClassName,
            )}
          />
          <div
            className={cn(
              "relative flex items-center justify-center rounded-full shadow-depth-outer",
              styles.rim,
              rimClassName,
            )}
          >
            <div
              className={cn(
                "relative aspect-[1/1] w-full overflow-hidden rounded-full",
                styles.inner,
                innerClassName,
              )}
            >
              {processedMedia}
              {children}
            </div>
          </div>
        </>
      ) : (
        <div
          className={cn(
            "relative aspect-[1/1] w-full overflow-hidden rounded-full shadow-depth-outer",
            styles.framelessSurface,
            surfaceClassName,
          )}
        >
          {processedMedia}
          {children}
        </div>
      )}
      {after}
    </Comp>
  );
}

AvatarFrame.displayName = "AvatarFrame";

export default AvatarFrame;
