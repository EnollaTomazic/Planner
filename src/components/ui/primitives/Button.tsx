"use client";

import * as React from "react";
import type { CSSProperties } from "react";
import { Slot } from "@radix-ui/react-slot";
import { motion, useReducedMotion } from "framer-motion";
import type { HTMLMotionProps } from "framer-motion";
import { cn, withBasePath } from "@/lib/utils";
import Spinner, { type SpinnerTone, type SpinnerSize } from "../feedback/Spinner";
import neumorphicStyles from "../neumorphic.module.css";
import { neuRaised, neuInset } from "./Neu";
import styles from "./Button.module.css";

export const buttonSizes = {
  sm: {
    height: "h-[var(--control-h-sm)]",
    padding: "px-[var(--space-4)]",
    text: "text-label",
    gap: "gap-[var(--space-1)]",
    icon: "[&_svg]:size-[var(--space-4)]",
  },
  md: {
    height: "h-[var(--control-h-md)]",
    padding: "px-[var(--space-4)]",
    text: "text-ui",
    gap: "gap-[var(--space-2)]",
    icon: "[&_svg]:size-[var(--space-5)]",
  },
  lg: {
    height: "h-[var(--control-h-lg)]",
    padding: "px-[var(--space-8)]",
    text: "text-title",
    gap: "gap-[var(--space-4)]",
    icon: "[&_svg]:size-[var(--space-8)]",
  },
  xl: {
    height: "h-[var(--control-h-xl)]",
    padding: "px-[calc(var(--control-h-xl)*4/3)]",
    text: "text-title-lg",
    gap: "gap-[calc(var(--control-h-xl)/3)]",
    icon: "[&_svg]:size-[calc(var(--control-h-xl)*4/3)]",
  },
} as const;

export type ButtonSize = keyof typeof buttonSizes;

type Tone = SpinnerTone;

const buttonSpinnerSizes: Record<ButtonSize, SpinnerSize> = {
  sm: "control-sm",
  md: "control-md",
  lg: "control-lg",
  xl: "control-xl",
};

const MotionSlot = motion.create(Slot);
const MotionAnchor = motion.a;

type AnchorMotionProps = HTMLMotionProps<"a">;

type ButtonBaseProps = {
  size?: ButtonSize;
  variant?: "primary" | "secondary" | "ghost";
  tone?: Tone;
  loading?: boolean;
  tactile?: boolean;
  glitch?: boolean;
};

type ButtonAsButtonProps = ButtonBaseProps &
  Omit<HTMLMotionProps<"button">, "ref" | "style"> & {
    asChild?: false;
    href?: undefined;
  };

type ButtonAsAnchorProps = ButtonBaseProps &
  Omit<AnchorMotionProps, "ref" | "style"> & {
    asChild?: false;
    href: AnchorMotionProps["href"];
  };

type ButtonAsChildProps = ButtonBaseProps &
  Omit<HTMLMotionProps<"button">, "ref" | "style"> & {
    asChild: true;
    href?: undefined;
  };

/**
 * Props for the {@link Button} component.
 * @property loading - When `true`, the button is disabled and `data-loading` is set.
 */
export type ButtonProps =
  | ButtonAsButtonProps
  | ButtonAsAnchorProps
  | ButtonAsChildProps;

export const colorVar: Record<Tone, string> = {
  primary: "--primary",
  accent: "--accent",
  info: "--accent-2",
  danger: "--danger",
};

const toneForegroundVar: Record<Tone, string> = {
  primary: "--primary-foreground",
  accent: "--accent-foreground",
  info: "--accent-2-foreground",
  danger: "--danger-foreground",
};

const primaryShadowVars: Record<Tone, string> = {
  primary:
    "[--glow-active:hsl(var(--primary)/0.35)] [--btn-primary-hover-shadow:0_var(--spacing-0-5)_calc(var(--space-3)/2)_calc(-1*var(--spacing-0-25))_hsl(var(--primary)/0.25)] [--btn-primary-active-shadow:inset_0_0_0_var(--spacing-0-25)_hsl(var(--primary)/0.6)]",
  accent:
    "[--glow-active:hsl(var(--accent)/0.35)] [--btn-primary-hover-shadow:0_var(--spacing-0-5)_calc(var(--space-3)/2)_calc(-1*var(--spacing-0-25))_hsl(var(--accent)/0.25)] [--btn-primary-active-shadow:inset_0_0_0_var(--spacing-0-25)_hsl(var(--accent)/0.6)]",
  info:
    "[--glow-active:hsl(var(--accent-2)/0.35)] [--btn-primary-hover-shadow:0_var(--spacing-0-5)_calc(var(--space-3)/2)_calc(-1*var(--spacing-0-25))_hsl(var(--accent-2)/0.25)] [--btn-primary-active-shadow:inset_0_0_0_var(--spacing-0-25)_hsl(var(--accent-2)/0.6)]",
  danger:
    "[--glow-active:hsl(var(--danger)/0.35)] [--btn-primary-hover-shadow:0_var(--spacing-0-5)_calc(var(--space-3)/2)_calc(-1*var(--spacing-0-25))_hsl(var(--danger)/0.25)] [--btn-primary-active-shadow:inset_0_0_0_var(--spacing-0-25)_hsl(var(--danger)/0.6)]",
};

const toneInteractionTokens: Record<Tone, string> = {
  primary: "[--hover:hsl(var(--primary)/0.14)] [--active:hsl(var(--primary)/0.2)]",
  accent: "[--hover:hsl(var(--accent)/0.14)] [--active:hsl(var(--accent)/0.2)]",
  info: "[--hover:hsl(var(--accent-2)/0.14)] [--active:hsl(var(--accent-2)/0.2)]",
  danger: "[--hover:hsl(var(--danger)/0.14)] [--active:hsl(var(--danger)/0.2)]",
};

const secondarySurfaceTokens: Record<Tone, string> = {
  primary: "[--hover:hsl(var(--primary)/0.25)] [--active:hsl(var(--primary)/0.35)]",
  accent: "[--hover:hsl(var(--accent)/0.25)] [--active:hsl(var(--accent)/0.2)]",
  info: "[--hover:hsl(var(--accent-2)/0.2)] [--active:hsl(var(--accent-2)/0.15)]",
  danger: "[--hover:hsl(var(--danger)/0.14)] [--active:hsl(var(--danger)/0.2)]",
};

const ghostSurfaceTokens: Record<Tone, string> = {
  primary: "[--hover:hsl(var(--background)/0.8)] [--active:hsl(var(--background))]",
  accent: "[--hover:hsl(var(--accent)/0.25)] [--active:hsl(var(--accent)/0.35)]",
  info: "[--hover:hsl(var(--accent-2)/0.25)] [--active:hsl(var(--accent-2)/0.35)]",
  danger: "[--hover:hsl(var(--danger)/0.1)] [--active:hsl(var(--danger)/0.2)]",
};

export const toneClasses: Record<
  NonNullable<ButtonProps["variant"]>,
  Record<Tone, string>
> = {
  primary: {
    primary: cn(toneInteractionTokens.primary, primaryShadowVars.primary),
    accent: cn(toneInteractionTokens.accent, primaryShadowVars.accent),
    info: cn(toneInteractionTokens.info, primaryShadowVars.info),
    danger: cn(toneInteractionTokens.danger, primaryShadowVars.danger),
  },
  secondary: {
    primary: cn(
      secondarySurfaceTokens.primary,
      "text-muted-foreground hover:text-foreground active:text-foreground focus-visible:text-foreground",
      "[--neu-surface:hsl(var(--panel)/0.6)]",
    ),
    accent: cn(
      secondarySurfaceTokens.accent,
      "text-on-accent",
      "[--neu-surface:hsl(var(--accent)/0.3)]",
    ),
    info: cn(
      secondarySurfaceTokens.info,
      "text-on-accent",
      "[--neu-surface:hsl(var(--accent-2)/0.25)]",
    ),
    danger: cn(
      secondarySurfaceTokens.danger,
      "text-[hsl(var(--danger-tint-foreground))]",
      "[--neu-surface:hsl(var(--danger)/0.25)]",
    ),
  },
  ghost: {
    primary: cn(
      ghostSurfaceTokens.primary,
      "text-foreground",
      "[--neu-surface:hsl(var(--card)/0.6)]",
      "border-[hsl(var(--line)/0.35)]",
    ),
    accent: cn(
      ghostSurfaceTokens.accent,
      "text-on-accent",
      "[--neu-surface:hsl(var(--accent)/0.2)]",
      "border-accent/35",
    ),
    info: cn(
      ghostSurfaceTokens.info,
      "text-on-accent",
      "[--neu-surface:hsl(var(--accent-2)/0.2)]",
      "border-accent-2/35",
    ),
    danger: cn(
      ghostSurfaceTokens.danger,
      "text-danger",
      "[--neu-surface:hsl(var(--danger)/0.12)]",
      "border-danger/35",
    ),
  },
};

type VariantConfigResult = {
  className: string;
  whileHover?: HTMLMotionProps<"button">["whileHover"];
  whileTap?: HTMLMotionProps<"button">["whileTap"];
  overlay?: React.ReactNode;
  contentClass?: string;
};

type VariantConfig = (options: {
  tone: Tone;
  tactile: boolean;
}) => VariantConfigResult;

export const variants: Record<
  NonNullable<ButtonProps["variant"]>,
  VariantConfig
> = {
  primary: ({ tone, tactile }) => ({
    className: cn(
      tactile
        ? "shadow-inner-md hover:shadow-inner-lg active:shadow-inner-lg"
        : "shadow-glow-md hover:shadow-btn-primary-hover active:shadow-btn-primary-active",
      tactile
        ? "active:translate-y-0"
        : "active:translate-y-[var(--spacing-0-25)]",
      tone === "primary"
        ? "[--neu-surface:hsl(var(--primary-soft))]"
        : `[--neu-surface:hsl(var(${colorVar[tone]})/0.12)]`,
      `border-[hsl(var(${colorVar[tone]})/0.35)]`,
      tone === "accent" || tone === "info"
        ? "text-on-accent"
        : `text-[hsl(var(${toneForegroundVar[tone]}))]`,
    ),
    whileTap: {
      scale: tactile ? 0.98 : 0.97,
    },
    contentClass:
      "relative z-10 inline-flex items-center gap-[var(--space-2)]",
  }),
  secondary: ({ tactile }) => ({
    className: cn(
      "[--neu-surface:hsl(var(--panel)/0.8)]",
      tactile
        ? "shadow-inner-md hover:shadow-inner-lg active:shadow-inner-lg"
        : "shadow-control hover:shadow-control-hover active:shadow-inner-md",
      tactile && "active:translate-y-0",
    ),
    whileHover: tactile
      ? { scale: 1.01 }
      : { scale: 1.02, boxShadow: neuRaised(15) },
    whileTap: tactile
      ? { scale: 0.98 }
      : {
          scale: 0.97,
          boxShadow: neuInset(9) as CSSProperties["boxShadow"],
        },
  }),
  ghost: () => ({
    className: "",
    whileTap: { scale: 0.97 },
  }),
} as const;

export const Button = React.forwardRef<
  HTMLButtonElement | HTMLAnchorElement,
  ButtonProps
>((props, ref) => {
  const {
    className,
    size = "md",
    variant = "secondary",
    tone = "primary",
    children,
    loading,
    tactile = false,
    glitch = false,
  } = props;
  const asChild = props.asChild ?? false;
  const reduceMotion = useReducedMotion();
  const disabledProp =
    "disabled" in props && typeof props.disabled !== "undefined"
      ? props.disabled
      : undefined;
  const isDisabled = Boolean(disabledProp) || Boolean(loading);
  const isLink =
    !asChild && "href" in props && typeof props.href !== "undefined";
  const toneColorVar = colorVar[tone];
  const s = buttonSizes[size];
  const spinnerSize = buttonSpinnerSizes[size];
  const base = cn(
    neumorphicStyles.neu,
    styles.root,
    styles.organicControl,
    glitch && "glitch-wrapper",
    glitch && styles.glitch,
    "relative inline-flex items-center justify-center rounded-[var(--control-radius)] border font-medium tracking-[0.02em] transition-all duration-quick ease-out motion-reduce:transition-none hover:bg-[--hover] active:bg-[--active] focus-visible:ring-2 focus-visible:ring-[var(--ring-contrast)] focus-visible:shadow-[var(--shadow-glow-md)] focus-visible:[outline:var(--spacing-0-5)_solid_var(--ring-contrast)] focus-visible:[outline-offset:var(--spacing-0-5)] disabled:opacity-disabled disabled:pointer-events-none data-[loading=true]:opacity-loading",
    "data-[disabled=true]:opacity-disabled data-[disabled=true]:pointer-events-none",
    "[--neu-radius:var(--control-radius)]",
    s.height,
    s.padding,
    s.text,
    s.gap,
    s.icon,
    className,
  );

  const {
    className: resolvedVariantClass,
    whileHover: variantHover,
    whileTap,
    overlay,
    contentClass,
  } = variants[variant]({ tone, tactile });

  const hoverAnimation = reduceMotion
    ? undefined
    : variant === "primary"
      ? { scale: tactile ? 1.02 : 1.03 }
      : variantHover;

  const contentClasses = cn(
    contentClass ?? cn("inline-flex items-center", s.gap),
    loading && "opacity-0",
  );

  const providedGlitchTextRaw = (props as Record<string, unknown>)["data-text"];
  const providedGlitchText =
    typeof providedGlitchTextRaw === "string" ? providedGlitchTextRaw : undefined;

  const glitchText = glitch
    ? providedGlitchText ??
        (typeof children === "string" ? children : undefined)
    : providedGlitchText;

  const mergedClassName = cn(
    base,
    resolvedVariantClass,
    toneClasses[variant][tone],
  );

  const renderInnerContent = (contentChildren: React.ReactNode) => (
    <>
      {variant === "primary" ? (
        <span
          className={cn(
            "absolute inset-0 pointer-events-none rounded-[inherit]",
            `bg-[linear-gradient(90deg,hsl(var(${toneColorVar})/.18),hsl(var(${toneColorVar})/.18))]`,
          )}
        />
      ) : (
        overlay
      )}
      {loading ? (
        <span className="pointer-events-none absolute inset-0 z-20 flex items-center justify-center">
          <Spinner size={spinnerSize} tone={tone} />
        </span>
      ) : null}
      <span className={contentClasses}>{contentChildren}</span>
    </>
  );

  if (asChild) {
    const slotProps = {
      ...(props as ButtonAsChildProps),
    } as Record<string, unknown> & {
      tabIndex?: number;
      type?: HTMLMotionProps<"button">["type"];
    };
    const tabIndex = slotProps.tabIndex;
    delete slotProps.tabIndex;
    delete slotProps.asChild;
    delete slotProps.size;
    delete slotProps.variant;
    delete slotProps.tone;
    delete slotProps.loading;
    delete slotProps.tactile;
    delete slotProps.glitch;
    delete slotProps.className;
    delete slotProps.children;
    delete slotProps.disabled;
    delete slotProps["data-text"];
    const baseProps = {
      className: mergedClassName,
      "data-loading": loading ? "true" : undefined,
      "data-disabled": isDisabled ? "true" : undefined,
      "data-tactile": tactile ? "true" : undefined,
      "data-glitch": glitch ? "true" : undefined,
      "data-text": glitchText,
      "aria-busy": loading ? true : undefined,
      whileHover: hoverAnimation,
      whileTap: reduceMotion ? undefined : whileTap,
      tabIndex: tabIndex ?? (isDisabled ? -1 : undefined),
      "data-variant": variant,
      "data-tone": tone,
      ...slotProps,
    };
    const childCount = React.Children.count(children);

    if (childCount !== 1 || !React.isValidElement(children)) {
      if (process.env.NODE_ENV !== "production") {
        console.warn(
          "[Button] `asChild` requires a single valid React element child.",
        );
      }

      return null;
    }
    const child = React.Children.only(
      children,
    ) as React.ReactElement<{ children?: React.ReactNode }>;

    return (
      <MotionSlot
        {...baseProps}
        ref={ref as React.ForwardedRef<HTMLElement>}
        aria-disabled={isDisabled ? true : undefined}
      >
        {React.cloneElement(
          child,
          undefined,
          renderInnerContent(child.props.children),
        )}
      </MotionSlot>
    );
  }

  if (isLink) {
    const anchorProps = {
      ...(props as ButtonAsAnchorProps),
    } as Record<string, unknown> & {
      href: ButtonAsAnchorProps["href"];
      target?: ButtonAsAnchorProps["target"];
      rel?: ButtonAsAnchorProps["rel"];
      download?: ButtonAsAnchorProps["download"];
      tabIndex?: number;
    };
    const href = anchorProps.href as ButtonAsAnchorProps["href"];
    const target = anchorProps.target as ButtonAsAnchorProps["target"];
    const rel = anchorProps.rel as ButtonAsAnchorProps["rel"];
    const download = anchorProps.download as ButtonAsAnchorProps["download"];
    const tabIndex = anchorProps.tabIndex;
    delete anchorProps.href;
    delete anchorProps.target;
    delete anchorProps.rel;
    delete anchorProps.download;
    delete anchorProps.tabIndex;
    delete anchorProps.asChild;
    delete anchorProps.size;
    delete anchorProps.variant;
    delete anchorProps.tone;
    delete anchorProps.loading;
    delete anchorProps.tactile;
    delete anchorProps.glitch;
    delete anchorProps.className;
    delete anchorProps.children;
    delete anchorProps["data-text"];
    const baseProps = {
      className: mergedClassName,
      "data-loading": loading ? "true" : undefined,
      "data-disabled": isDisabled ? "true" : undefined,
      "data-tactile": tactile ? "true" : undefined,
      "data-glitch": glitch ? "true" : undefined,
      "data-text": glitchText,
      "aria-busy": loading ? true : undefined,
      whileHover: hoverAnimation,
      whileTap: reduceMotion ? undefined : whileTap,
      tabIndex: tabIndex ?? (isDisabled ? -1 : undefined),
      "data-variant": variant,
      "data-tone": tone,
      ...anchorProps,
    };
    let resolvedHref = href;

    const resolvedRel =
      target === "_blank" && typeof rel === "undefined"
        ? "noopener noreferrer"
        : rel;

    if (typeof href === "string") {
      const trimmedHref = href.trim();
      const hasScheme = /^[a-zA-Z][a-zA-Z\d+.-]*:/.test(trimmedHref);
      const isProtocolRelative = trimmedHref.startsWith("//");
      const isHash = trimmedHref.startsWith("#");
      const shouldPrefixBasePath =
        !isHash &&
        (trimmedHref.startsWith("/") || (!hasScheme && !isProtocolRelative));

      resolvedHref = shouldPrefixBasePath
        ? withBasePath(trimmedHref)
        : trimmedHref;
    }

    return (
      <MotionAnchor
        {...baseProps}
        ref={ref as React.ForwardedRef<HTMLAnchorElement>}
        href={resolvedHref}
        target={target}
        rel={resolvedRel}
        download={download}
        aria-disabled={isDisabled ? true : undefined}
      >
        {renderInnerContent(children as React.ReactNode)}
      </MotionAnchor>
    );
  }

  const buttonProps = {
    ...(props as ButtonAsButtonProps),
  } as Record<string, unknown> & {
    type?: HTMLMotionProps<"button">["type"];
    tabIndex?: number;
  };
  const typeProp =
    (buttonProps.type as HTMLMotionProps<"button">["type"]) ?? "button";
  const tabIndex = buttonProps.tabIndex;
  delete buttonProps.type;
  delete buttonProps.tabIndex;
  delete buttonProps.asChild;
  delete buttonProps.size;
    delete buttonProps.variant;
  delete buttonProps.tone;
  delete buttonProps.loading;
  delete buttonProps.tactile;
  delete buttonProps.glitch;
  delete buttonProps.className;
  delete buttonProps.children;
  delete buttonProps["data-text"];
  delete buttonProps.disabled;
  const baseProps = {
    className: mergedClassName,
    "data-loading": loading ? "true" : undefined,
    "data-disabled": isDisabled ? "true" : undefined,
    "data-tactile": tactile ? "true" : undefined,
    "data-glitch": glitch ? "true" : undefined,
    "data-text": glitchText,
    "aria-busy": loading ? true : undefined,
    whileHover: hoverAnimation,
    whileTap: reduceMotion ? undefined : whileTap,
    tabIndex,
    "data-variant": variant,
    "data-tone": tone,
    ...buttonProps,
  };

  return (
    <motion.button
      {...baseProps}
      ref={ref as React.ForwardedRef<HTMLButtonElement>}
      type={typeProp}
      disabled={isDisabled}
    >
      {renderInnerContent(children as React.ReactNode)}
    </motion.button>
  );
});

Button.displayName = "Button";
export default Button;
