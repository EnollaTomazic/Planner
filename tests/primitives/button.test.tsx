import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import { describe, it, expect, afterEach, vi } from "vitest";
import { Button } from "../../src/components/ui/primitives/Button";

afterEach(cleanup);

describe("Button", () => {
  it("renders its children", () => {
    const { getByRole } = render(<Button>Click me</Button>);
    expect(getByRole("button")).toHaveTextContent("Click me");
  });

  it("applies variant classes", () => {
    const { getByRole } = render(
      <Button className="btn-primary">Click me</Button>,
    );
    expect(getByRole("button")).toHaveClass("btn-primary");
  });

  it("has no outline when focused", () => {
    const { getByRole } = render(<Button>Focus</Button>);
    const btn = getByRole("button");
    btn.focus();
    const style = getComputedStyle(btn);
    expect(style.outlineStyle === "none" || style.outlineStyle === "").toBe(true);
    expect(style.outlineWidth === "0px" || style.outlineWidth === "").toBe(true);
  });

  it("has reduced opacity and no pointer events when disabled", () => {
    const onClick = vi.fn();
    const { getByRole } = render(
      <Button disabled onClick={onClick}>
        Disabled
      </Button>,
    );
    const btn = getByRole("button");
    expect(btn).toHaveClass(
      "disabled:opacity-50",
      "disabled:pointer-events-none",
    );
    fireEvent.click(btn);
    expect(onClick).not.toHaveBeenCalled();
  });

  it.each([
    ["sm", "[&>svg]:size-4"],
    ["md", "[&>svg]:size-5"],
    ["lg", "[&>svg]:size-6"],
  ])("applies %s icon sizing", (size, cls) => {
    const { getByRole } = render(
      <Button size={size as any}>
        <svg />
      </Button>,
    );
    expect(getByRole("button")).toHaveClass(cls);
  });

  it.each([
    ["sm", "gap-1"],
    ["md", "gap-2"],
    ["lg", "gap-3"],
  ])("applies %s gap spacing", (size, cls) => {
    const { getByRole } = render(
      <Button size={size as any}>
        <svg />
        Label
      </Button>,
    );
    expect(getByRole("button")).toHaveClass(cls);
  });
});
