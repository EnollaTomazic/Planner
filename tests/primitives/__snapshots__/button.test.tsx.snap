// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > matches snapshot for secondary danger tone 1`] = `
<button
  class="relative inline-flex items-center justify-center rounded-[var(--control-radius)] border font-medium tracking-[0.02em] transition-all duration-[var(--dur-quick)] ease-out motion-reduce:transition-none hover:bg-[--hover] active:bg-[--active] focus-visible:[outline:none] focus-visible:ring-2 focus-visible:ring-[var(--focus)] disabled:opacity-[var(--disabled)] disabled:pointer-events-none data-[loading=true]:opacity-[var(--loading)] data-[disabled=true]:opacity-[var(--disabled)] data-[disabled=true]:pointer-events-none h-[var(--control-h-md)] px-[var(--space-4)] text-ui gap-[var(--space-2)] [&_svg]:size-[var(--space-5)] shadow-[var(--btn-secondary-shadow-rest)] hover:shadow-[var(--btn-secondary-shadow-hover)] active:shadow-[var(--btn-secondary-shadow-active)] [--hover:theme('colors.interaction.danger.hover')] [--active:theme('colors.interaction.danger.active')] text-danger-foreground bg-danger/25"
  style="--btn-secondary-shadow-rest: var(--space-3) var(--space-3) var(--space-5) hsl(var(--panel)/0.72), calc(var(--space-3) * -1) calc(var(--space-3) * -1) var(--space-5) hsl(var(--foreground)/0.06); --btn-secondary-shadow-hover: calc(var(--space-1) * 3.75) calc(var(--space-1) * 3.75) calc(var(--space-1) * 7.5) hsl(var(--panel)/0.72), calc(var(--space-1) * -3.75) calc(var(--space-1) * -3.75) calc(var(--space-1) * 7.5) hsl(var(--foreground)/0.06); --btn-secondary-shadow-active: inset calc(var(--space-1) * 1.5) calc(var(--space-1) * 1.5) calc(var(--space-1) * 3.5) hsl(var(--panel)/0.85), inset calc(var(--space-1) * -1.5) calc(var(--space-1) * -1.5) calc(var(--space-1) * 3.5) hsl(var(--foreground)/0.08);"
  tabindex="0"
  type="button"
>
  <span
    class="inline-flex items-center gap-[var(--space-2)]"
  >
    Delete
  </span>
</button>
`;
